### PR TITLE
Initial workflow implementation

### PR DESCRIPTION
**Pull request by ahgittin**
Essential structure and sashimi slice of new declarative YAML workflow, per discussion on ML and proposal https://docs.google.com/document/d/1u02Bi6sS8Fkf1s7UzRRMnvLhA477bqcyxGa0nJesqkI

Includes the following:
* no-op and sleep steps
* workflow-effector initializer
* creation and execution of tasks from step definitions
* java and camp yaml tests

many more step types and workflow functionality to be added, but this provides a base where the essential definitions work and subsequent tasks can be done in parallel

### CODE REVIEW

**Comment by algairim** [comment]
Good start. Merging it.

**Review by geomacy — COMMENTED** [review]
@ahgittin this looks great. My main impression is one of surprise as to how simple it is, I guess I had in mind that it would take more work do do all this. Well, it's simple so far, I guess. No doubt it will get more complex as later aspects get added, like traceability and recovery, and error handling.

Added a couple of minor comments or questions. The one on the duplicate test step is probably worth an update.

File: core/src/main/java/org/apache/brooklyn/core/workflow/steps/SetSensorWorkflowStep.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    @Override
+    public void setShorthand(String expression) {
+        this.sensor = EntityValueToSet.parseFromShorthand(expression, this::setValue);
+    }
+
+    @Override
+    protected Task<?> newTask(String name, WorkflowExecutionContext workflowExecutionContext) {
+        return Tasks.create(getDefaultTaskName(workflowExecutionContext), () -> {
```
[the remark this answers is not in the document]
Comment (ahgittin) [reply core/src/main/java/org/apache/brooklyn/core/workflow/steps/SetSensorWorkflowStep.java:59]: good spot, needs tidying.  the `name` of the current step is now available in the `workflowExecutionContext`, and included in some cases, so i think not needed on this api.

File: camp/camp-brooklyn/src/test/java/org/apache/brooklyn/camp/brooklyn/WorkflowYamlTest.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+                "services:",
+                "- type: " + BasicEntity.class.getName(),
+                "  brooklyn.initializers:",
+                "  - type: workflow-effector",
+                "    brooklyn.config:",
+                "      name: myWorkflow",
+                "      steps:",
+                "        step1:",
+                "          type: no-op",
+                "        step1:",
```
[the remark this answers is not in the document]
Comment (ahgittin) [reply camp/camp-brooklyn/src/test/java/org/apache/brooklyn/camp/brooklyn/WorkflowYamlTest.java:98]: yaml removes the duplicates so it is quite low level.  actually a good argument for switching to a list-based definition.

File: core/src/main/java/org/apache/brooklyn/core/workflow/steps/SetSensorWorkflowStep.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (ahgittin) [reply core/src/main/java/org/apache/brooklyn/core/workflow/steps/SetSensorWorkflowStep.java:59]: signature change in #1361 at https://github.com/apache/brooklyn-server/pull/1361/files#diff-061e4e878f0722d9f76ae09ddde336a8ead2d9982e7459125033ab1c594ba987R92

**Pull request merged**